### PR TITLE
Plane: Onion part 1b

### PR DESCRIPTION
**Pull request by magicrub**
This replaces https://github.com/ArduPilot/ardupilot/pull/10107 by rolling back a couple of the changes that held mode properties. It should be more "obviously correct" with structure-only changes. Next part will be where we are doing more fancy things with the mode classes. Let's get the structure in place and merged into master so we can then more easily discuss how to do all the optimization stuff with using the mode classes.

I marked this WIP because I plan to break the commits up a little better and add some comments. No source code changes are planned. This PR is here to get more eyes on this and get some testing happening. It would be a HUGE help if @Naterater could do some flight tests with this in all the flight modes. Ping us if you need a binary!

### CODE REVIEW

File: ArduPlane/mode_acro.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#include "mode.h"
+#include "Plane.h"
+
+bool ModeAcro::_enter()
```
[the remark this answers is not in the document]
Comment (magicrub) [reply ArduPlane/mode_acro.cpp]: ... not when we're done with the onion!

[the remark this answers is not in the document]
Comment (tridge) [reply ArduPlane/mode_acro.cpp]: we do fail to enter Q modes actually, though it works by switching back to previous_mode

File: ArduPlane/avoidance_adsb.h [hunk 2]
```diff
@@ -34,5 +34,5 @@ class AP_Avoidance_Plane : public AP_Avoidance {
     bool handle_avoidance_horizontal(const AP_Avoidance::Obstacle *obstacle, bool allow_mode_change);
 
     // control mode before avoidance began
-    FlightMode prev_control_mode = RTL;
+    enum Mode::Number prev_control_mode_number = Mode::Number::RTL;
```
[the remark this answers is not in the document]
Comment (magicrub) [reply ArduPlane/avoidance_adsb.h]: I could, I guess I was just trying to keep it simple. I'll take a look at making it an object

**Comment by magicrub** [comment]
[arduplane.zip](https://github.com/ArduPilot/ardupilot/files/2725584/arduplane.zip)
I've attached the apj file but renamed it to zip to post here. Git hash is a81772cb34

**Comment by amilcarlucas** [comment]
For reference, the Flight modes are defined in Mavlink enums:
https://mavlink.io/en/messages/ardupilotmega.html#COPTER_MODE

**Comment by magicrub** [comment]
@amilcarlucas For [Plane modes](https://mavlink.io/en/messages/ardupilotmega.html#PLANE_MODE), we have [this](https://github.com/magicrub/ardupilot/blob/plane_onion3_flat/ArduPlane/mode.h#L15). Just like the other vehicles.

**Comment by magicrub** [comment]
rebased on master and updated commit msg

**Comment by magicrub** [comment]
@amilcarlucas fair request. However, all vehicles need that change. Let's leave that for another PR.

**Comment by amilcarlucas** [comment]
@magicrub OK, I'll do a PR once this one goes in.

**Comment by OXINARF** [comment]
@magicrub @amilcarlucas We've been craving for an AP_Telemetry library forever, tying things more into MAVLink only makes that job harder, not easier... The flight mode numbers have its origin in ArduPilot, not in MAVLink.

**Comment by peterbarker** [comment]
On Tue, 15 Jan 2019, Amilcar Lucas wrote:

> @magicrub OK, I'll do a PR once this one goes in.

You could make an equivalent PR against Copter or Rover so we can consider 
:-)

**Comment by rmackay9** [comment]
@magicrub, I should just stay quiet but I'm excited about this PR.. I guess you said you wanted to do some more testing or are you happy for this to go in now?  I see the WIP label is gone..

**Comment by amilcarlucas** [comment]
@OXINARF we can cross that bridge once we get there :)
And we can at that point, replace the assignments with static_asserts.

**Comment by amilcarlucas** [comment]
Needs a rebase

**Comment by magicrub** [comment]
rebased

**Comment by Naterater** [comment]
Hey @magicrub, what's the status of this?

**Comment by magicrub** [comment]
I just rebased this. @peterbarker and I will be flying it here in Canberra dev conference. Plan is to pound on it and get it merged this weekend

**Comment by magicrub** [comment]
This was just flown on two aircraft at CMAC, owned by @peterbarker and @tridge . So, things are looking promising for a merge very soon!

File: ArduPlane/avoidance_adsb.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (magicrub) [reply ArduPlane/avoidance_adsb.h]: fixed

**Comment by magicrub** [comment]
rebased. I've also added the DevTopic tag. Let's leave that tag on until this merges

**Comment by magicrub** [comment]
This needs some eyeballs on it. Can a few other devs PLEASE take a look at this?

File: ArduPlane/system.cpp [hunk 3]
```diff
@@ -318,200 +289,39 @@ void Plane::set_mode(enum FlightMode mode, mode_reason_t reason)
...
-        auto_navigation_mode = false;
-        if (!quadplane.init_mode()) {
-            control_mode = previous_mode;
-        } else {
-            auto_throttle_mode = false;
-            auto_state.vtol_mode = true;
-        }
-        break;
+        // we failed entering new mode, roll back to old
+        previous_mode = &old_previous_mode;
```
Comment (WickedShell) [inline ArduPlane/system.cpp]: Don't you need to re-enter the old mode, as enter a mode has significant side effects? (IE navigation target etc).

Comment (magicrub) [reply ArduPlane/system.cpp]: yeah, quadplane is the black duck here. It's the only thing that can fail a set_mode. I don't have a quadplane to test thoroughly. Can you test it for me?

Comment (rmackay9) [reply ArduPlane/system.cpp]: @magicrub, I'd vote for testing it in SITL..

**Comment by magicrub** [comment]
I think @tridge 's baby hit the merge button on accident!

**Pull request merged**